### ISSUE TITLE
Factory class method provided by a mixin can define a duplicated argument.

### ISSUE THREAD

**Issue by cainmagi** [microsoft/pyright#8964]
## Describe the bug

A mixin class can define a factory method returning the instance of the subclass. However, the definition of the factory method can contain extra arguments that may be duplicated in the initialization method of the subclass. Currently, pyright does not report the correct behavior in this case.

## Code or Screenshots

Test the demo here:
https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIANFAMLkA2D5ARg8TQAqXkQDKCYgGMaIYgDdijAPrxBAKHmcoAXijcQvAcIAUAIk56AlPIKrCJCiH0Fji%2BUOYBnJ1ACySAB6oAXPKgBUAACjuQuEMQwABZgACb%2BgbHEwFDA5EIwuHA6jk4%2BdIzMbMQA2pw0BAC6NLHkMOT5TjDUUABUlGh56gB0HU40ra0A1gDuffmc3aN9RlAAtAB8hH6Bq1AIIKgwOmIbOrX1RqLEezPHe9MgnUazUADEUADkMMSeMI9QOrMA3gC%2BCWsoGIYABXEAoKC5HTtK79NrDMawkyKUIuKAAOVe2w83hQRhWiWSUGk0lQSBgJJ0TmIDGANTqDUwKHwagAjLNFhiwChiATAQgwk57GJJDI5MQdJi3t00hksrcHkRBFAwCk9FKYDL0pkQHA9JhXHp9gzGs16fV8ltzOz5ksNXp5OJGOYNVq5brjfUVHoXm87CKpAxZCQdE6GAqMVj5EA

Codes:

```python
from typing import TypeVar, Callable, ParamSpec, reveal_type

P = ParamSpec("P")
T = TypeVar("T")


class Mixin:
    @classmethod
    def factory(cls: Callable[P, T], data: str, *args: P.args, **kwargs: P.kwargs) -> T:
        print(repr(data), repr(args), repr(kwargs))  # 'text' () {}
        return cls(*args, **kwargs)


class Next(Mixin):
    def __init__(self, data: int = 1) -> None:
        pass


reveal_type(Next.factory)  # Type of "Next.factory" is "(data: str, data: int = 1) -> Next"
val = Next.factory(data="text")
reveal_type(val)  # Next
```

During the run time, the above codes can run successfully, where the argument `data` is captured by `Mixin` and `Next` gets `data=1`. However, pyright reports the signature of `Next.factory` as:
```python
Next.factory(data: str, data: int = 1) -> Next
```
However, I expect that this case should be
```python
Next.factory(data: str) -> Next
```

## VS Code extension or command-line

* `pyright`: 1.1.378
* `pylance`: 2024.9.1
* `python`: 3.12.2.final.0
* `VSCode`: 1.93.0
* `OS`: Windows x64 10.0.22631

**Comment by erictraut** [comment]
I'm not sure I understand what problem you're seeing here. 

Yes, it's possible for the specialization of a `ParamSpec` to result in a duplicate parameter name within a signature. This condition that would be illegal if you attempted to define such a signature in a `def` statement, but that's not what you're doing here.

It would be incorrect to drop the second parameter as you're suggesting because it's legal to call the `Next.factory` with two positional parameters such as `Next.factory("", 0)`.

Not surprisingly, mypy has the same behavior here.

**Comment by cainmagi** [comment]
Pyright does not have exactly the same behavoir of mypy. If I use no arguments, pyright will **NOT** raise any issue but mypy **DOES**:
```python
reveal_type(Next.factory)
val = Next.factory()  # Should raise an issue here.
reveal_type(val)
```
See
* pyright: https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIANFAMLkA2D5ARg8TQAqXkQDKCYgGMaIYgDdijAPrxBAKHmcoAXijcQvAcIAUAIk56AlPIKrCJCiH0Fji%2BUOYBnJ1ACySAB6oAXPKgBUAACjuQuEMQwABZgACb%2BgbHEwFDA5EIwuHA6jk4%2BdIzMbMQA2pw0BAC6NLHkMOT5TjDUUABUlGh56gB0HU40ra0A1gDuffmc3aN9RlAAtAB8hH6Bq1AIIKgwOmIbOrX1RqLEezPHe9MgnUazUADEUADkMMSeMI9QOrMA3gC%2BCWsoGIYABXEAoKC5HTtK79NrDMawkyKUIuKAAOVe2w83hQRhWiWSUGk0lQSBgJJ0TmIDGANTqDUwKHwagAjLNFhiwChiATAQgwk57GJJDI5MQdJi3t00hksiZxIxzFKYDL0pkQNlbg8%2BDEQQxYkDyEhqVByBCTU4QcQoFFiGJuvIRVIGLISDpFQwTEA
* mypy: https://mypy-play.net/?mypy=latest&python=3.12&gist=76ad53f41fde338efacdbd2d12f521e2

**Comment by cainmagi** [comment]
Another example is to change the definition of `Mixin` by a little bit. In this case, pyright **DOES** reports an issue but mypy **DOES NOT**.
```python
from typing import TypeVar, Callable, ParamSpec, reveal_type

P = ParamSpec("P")
T = TypeVar("T")


class Mixin:
    @classmethod
    def factory(
        cls: Callable[P, T], data: str, /, *args: P.args, **kwargs: P.kwargs  # Make data purely positional
    ) -> T:
        print(repr(data), repr(args), repr(kwargs))  # '' () {'data': 2}
        return cls(*args, **kwargs)


class Next(Mixin):
    def __init__(self, data: int = 1) -> None:
        pass


reveal_type(Next.factory)
val = Next.factory("", data=2)  # Parameter "data" is already assigned  (reportCallIssue)
reveal_type(val)
```

See
* pyright: https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIANFAMLkA2D5ARg8TQAqXkQDKCYgGMaIYgDdijAPrxBAKHmcoAXijcQvAcIAUAIk56AlPIKrCJCiH0Fji%2BUOYBnJ1ACySAB6oAXPKgBUAACjuQuEMQwABZgACb%2BgbHEwFDA5EIwuHA6CYGBjk4%2BdIzMbMQA2pw0BAC6NLHkMORFTjDUUAD0NABUlGiF6gB0fU493QDWAO4jRZyDUyMBAMTu5OPEUA1NUAgArmIMcDtgTkgwSGAojLlQRlAAtAB8hH55eQggqDA6Yh86W%2BQjKJiH8RkCoL9rAsQP0jHcoCsAOSIqA6O4Ab0RAMRRQATABfG55MQwfYoKAFHS9GGjKDdCbTGkmRShFxQAByxE83w83hQRleiWSUGk0lQZ1FOicxAYwHqjWamBQ%2BDUAEY7k8OZdiIK3jswk57GJJDI5MQdJzuYM0hksiZxIxzJaYNb0pkQNk9Hp5U0VLj4SsNLxIsQQFA9AC9JhXIwxORYkcDUg0ChiLEAj8QbgYPQmABJFy7YgmY1SBiyEg6B0MExAA
* mypy: https://mypy-play.net/?mypy=latest&python=3.12&gist=e4d8230e9ae75cd5f21a56be54cb175a

**Comment by erictraut** [comment]
Yes, I agree those are bugs.

Pyright's logic for argument matching was written prior to the introduction of `ParamSpec`, and it wasn't updated to handle the situation where a parameter name appears more than once.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.382.

**Comment by cainmagi** [comment]
> Pyright does not have exactly the same behavoir of mypy. If I use no arguments, pyright will **NOT** raise any issue but mypy **DOES**:
> 
> ```python
> reveal_type(Next.factory)
> val = Next.factory()  # Should raise an issue here.
> reveal_type(val)
> ```
> 
> See
> 
> * pyright: https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIANFAMLkA2D5ARg8TQAqXkQDKCYgGMaIYgDdijAPrxBAKHmcoAXijcQvAcIAUAIk56AlPIKrCJCiH0Fji%2BUOYBnJ1ACySAB6oAXPKgBUAACjuQuEMQwABZgACb%2BgbHEwFDA5EIwuHA6jk4%2BdIzMbMQA2pw0BAC6NLHkMOT5TjDUUABUlGh56gB0HU40ra0A1gDuffmc3aN9RlAAtAB8hH6Bq1AIIKgwOmIbOrX1RqLEezPHe9MgnUazUADEUADkMMSeMI9QOrMA3gC%2BCWsoGIYABXEAoKC5HTtK79NrDMawkyKUIuKAAOVe2w83hQRhWiWSUGk0lQSBgJJ0TmIDGANTqDUwKHwagAjLNFhiwChiATAQgwk57GJJDI5MQdJi3t00hksiZxIxzFKYDL0pkQNlbg8%2BDEQQxYkDyEhqVByBCTU4QcQoFFiGJuvIRVIGLISDpFQwTEA
> * mypy: https://mypy-play.net/?mypy=latest&python=3.12&gist=76ad53f41fde338efacdbd2d12f521e2

Hello, Eric!

Thank you for updating this! I have checked the examples just now. The positional argument issue has been perfectly solved. But I can see that `pyright==1.1.382` still does not raise an issue for this example. I can confirm that this usage will raise an error in run time, so I think `pyright` supposes to raise an issue like `mypy` does. Do you think so? Please let me know your opinion. Thank you!
